src/frontend/parser.mly

```diff
@@ -218,61 +232,125 @@ unsized_dims:
+          loc= Location_span.of_positions_exn $startpos $endpos
+        }
+    })
+    }
+  (* Array dimensions option must be inlined, else it will conflict with first
+     rule. *)
+  | dims_opt=ioption(arr_dims) ty=type_rule
+      id_rhs=id_and_optional_assignment(rhs) SEMICOLON
+    { (fun ~is_global ->
+      (* map over each variable in v (often only one), assigning each the same
+         type. *)
+      let dims = Option.value dims_opt ~default:[] in
+      let (id, rhs_opt) = id_rhs in
+          { stmt=
+              VarDecl {
+                  decl_type= Sized (reducearray (fst ty, dims))
+                ; transformation= snd ty
+                ; identifier= id
+                ; initial_value= rhs_opt
+                ; is_global
+                }
+          ; smeta= {
+              loc=
+                (* From the docs:
+                We remark that, if the current production has an empty right-hand side,
+                then $startpos and $endpos are equal, and (by convention) are the end
+                position of the most recently parsed symbol (that is, the symbol that
+                happens to be on top of the automaton’s stack when this production is
+                reduced). If the current production has a nonempty right-hand side,
+                then $startpos is the same as $startpos($1) and $endpos is the same
+                as $endpos($n), where n is the length of the right-hand side.
+
+
+                So when dims_opt is empty, it uses the preview token as its startpos,
+                but that makes the whole declaration think it starts at the previous
+                token
+                 *)
+                let startpos = match dims_opt with
+                  | None -> $startpos(ty)
+                  | Some _ -> $startpos
+                in
+                Location_span.of_positions_exn startpos $endpos
+            }
+          }
+    )}
+
 var_decl:
-  | sbt=sized_basic_type id=decl_identifier d=option(dims)
-    ae=option(pair(ASSIGN, expression)) SEMICOLON
+  | d_fn=decl(sized_basic_type, expression)
     { grammar_logger "var_decl" ;
-      let sizes = match d with None -> [] | Some l -> l in
-      {stmt=
-         VarDecl {decl_type= Sized (reducearray (sbt, sizes));
-                  transformation= Identity;
-                  identifier= id;
-                  initial_value=Option.map ~f:snd ae;
-                  is_global= false};
-       smeta= {loc = Location_span.of_positions_exn $startpos $endpos}}
+      d_fn ~is_global:false
+    }
+
+top_var_decl:
+  | d_fn=decl(top_var_type, expression)
+    { grammar_logger "top_var_decl" ;
+      d_fn ~is_global:true
+    }
+
+top_var_decl_no_assign:
+  | d_fn=decl(top_var_type, no_assign)
+    { grammar_logger "top_var_decl_no_assign" ;
+      d_fn ~is_global:true
     }
 
 sized_basic_type:
   | INT
-    { grammar_logger "INT_var_type" ; SizedType.SInt }
+    { grammar_logger "INT_var_type" ; (SizedType.SInt, Identity) }
```

Comment:
What are these `, Identity` used for? Having trouble tracking it down.

Reply:
They're transformations, they're set to non-identity in e.g. top_var_type